zebra-network/src/peer_set/candidate_set.rs

```diff
@@ -329,24 +332,73 @@ where
 /// If the data in an address is invalid, this function can:
 /// - modify the address data, or
 /// - delete the address.
-//
-// TODO: re-enable this lint when last_seen_limit is used
-#[allow(unused_variables)]
+///
+/// # Security
+///
+/// Adjusts untrusted last seen times so they are not in the future. This stops
+/// malicious peers keeping all their addresses at the front of the connection
+/// queue. Honest peers with future clock skew also get adjusted.
+///
+/// Rejects all addresses if there are at least two that have reported
+/// last_seen` times where one is so far in the future and another is so far in
+/// the past that they cause an overflow when offsetting the times.
 fn validate_addrs(
     addrs: impl IntoIterator<Item = MetaAddr>,
     last_seen_limit: DateTime<Utc>,
-) -> impl IntoIterator<Item = MetaAddr> {
+) -> impl Iterator<Item = MetaAddr> {
     // Note: The address book handles duplicate addresses internally,
     // so we don't need to de-duplicate addresses here.
 
     // TODO:
     // We should eventually implement these checks in this function:
-    // - Zebra should stop believing far-future last_seen times from peers (#1871)
     // - Zebra should ignore peers that are older than 3 weeks (part of #1865)
     //   - Zebra should count back 3 weeks from the newest peer timestamp sent
     //     by the other peer, to compensate for clock skew
     // - Zebra should limit the number of addresses it uses from a single Addrs
     //   response (#1869)
 
-    addrs
+    let mut addrs: Vec<_> = addrs.into_iter().collect();
+
+    if !addrs.is_empty() {
+        limit_last_seen_times(&mut addrs, last_seen_limit);
+    }
+
+    addrs.into_iter()
+}
+
+/// Ensure all reported `last_seen` times are less than or equal to `last_seen_limit`.
+///
+/// This will consider all addresses as invalid if trying to offset their
+/// `last_seen` times to be before the limit causes an overflow.
+///
+/// # Panics
+///
+/// If the `addrs` list is empty.
+fn limit_last_seen_times(addrs: &mut Vec<MetaAddr>, last_seen_limit: DateTime<Utc>) {
+    let (oldest_reported_seen_time, newest_reported_seen_time) = addrs.iter().fold(
+        (chrono::MAX_DATETIME, chrono::MIN_DATETIME),
+        |(oldest, newest), addr| {
+            let last_seen = addr.get_last_seen();
+            (oldest.min(last_seen), newest.max(last_seen))
+        },
+    );
+
+    // If any time is in the future, adjust all times, to compensate for clock skew on honest peers
+    if newest_reported_seen_time > last_seen_limit {
+        let offset = last_seen_limit - newest_reported_seen_time;
+
+        // Check if applying the offset can cause an overflow
+        if oldest_reported_seen_time
+            .checked_add_signed(offset)
+            .is_some()
```

Comment:
Yeah, I'm not sure how to handle this. A few options come to mind:

1. Leave as is
2. Move the logic from `MetaAddr::offset_last_seen_by` into here, and add `MetaAddr::set_last_seen` setter
3. Rewrite this so that there is no initial check, but `MetaAddr::offset_last_seen_by` returns something (`bool`? `Result<(), OverflowError>`?)

Should we replace the `MetaAddr::offset_last_seen_by` with just `MetaAddr::set_last_seen`?

Reply:
I think it would be better to have all the time calculations in one place. So when we make changes we can see all the impacted code.

So a `MetaAddr::set_last_seen` method is the way to go. We could make it `pub(crate)` so it's clearly an internal use function.

(Zebra mostly does comparisons on times. There's one other time calculation in block validation, which adds 2 hours to the current time. So I think this is our biggest set of time calculations.)